File: scripts/remove_bg.py
"""Remove dark background from Makima PNG images using color-distance thresholding."""
from PIL import Image
import os

IMG_DIR = os.path.join(os.path.dirname(__file__), '..', 'assets', 'images')
BG_COLOR = (26, 26, 26)    # target dark background
THRESHOLD = 55              # max color distance to be considered background
FEATHER = 15                # feather zone: partial alpha transition

def color_dist(c1, c2):
    return abs(c1[0]-c2[0]) + abs(c1[1]-c2[1]) + abs(c1[2]-c2[2])

def process_image(path):
    img = Image.open(path).convert('RGBA')
    w, h = img.size
    pixels = img.load()

    # Pass 1: mark background pixels
    alpha = [[0]*h for _ in range(w)]
    for y in range(h):
        for x in range(w):
            r, g, b, _ = pixels[x, y]
            d = color_dist((r, g, b), BG_COLOR)
            if d <= THRESHOLD:
                alpha[x][y] = 0
            elif d <= THRESHOLD + FEATHER:
                # Feather: linear transition from 0 to 255
                alpha[x][y] = int(255 * (d - THRESHOLD) / FEATHER)
            else:
                alpha[x][y] = 255

    # Apply alpha
    for y in range(h):
        for x in range(w):
            r, g, b, a = pixels[x, y]
            pixels[x, y] = (r, g, b, alpha[x][y])

    img.save(path)
    print(f'  {os.path.basename(path)}  done')

def main():
    files = sorted(f for f in os.listdir(IMG_DIR) if f.startswith('makima_') and f.endswith('.png'))
    print(f'Processing {len(files)} images...')
    for f in files:
        process_image(os.path.join(IMG_DIR, f))
    print('All done.')

if __name__ == '__main__':
    main()

File: scripts/remove_bg.mjs
import sharp from 'sharp';
import { readdir, readFile } from 'fs/promises';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const IMG_DIR = join(__dirname, '..', 'assets', 'images');

const BG = [26, 26, 26];       // target dark background RGB
const THRESHOLD = 55;           // max color distance for full transparency
const FEATHER = 15;             // transition zone

function dist(r, g, b) {
  return Math.abs(r - BG[0]) + Math.abs(g - BG[1]) + Math.abs(b - BG[2]);
}

async function processImage(filePath) {
  const { data, info } = await sharp(filePath)
    .ensureAlpha()
    .raw()
    .toBuffer({ resolveWithObject: true });

  const { width, height, channels } = info; // channels = 4 (RGBA)
  const pixels = new Uint8Array(data);

  for (let i = 0; i < pixels.length; i += 4) {
    const r = pixels[i];
    const g = pixels[i + 1];
    const b = pixels[i + 2];
    const d = dist(r, g, b);

    if (d <= THRESHOLD) {
      pixels[i + 3] = 0; // fully transparent
    } else if (d <= THRESHOLD + FEATHER) {
      pixels[i + 3] = Math.round(255 * (d - THRESHOLD) / FEATHER);
    }
    // else: keep existing alpha (255)
  }

  await sharp(pixels, { raw: { width, height, channels } })
    .png()
    .toFile(filePath + '.tmp');

  // Replace original
  const { rename } = await import('fs/promises');
  await rename(filePath + '.tmp', filePath);
}

async function main() {
  const files = (await readdir(IMG_DIR))
    .filter(f => f.startsWith('makima_') && f.endsWith('.png'))
    .sort();

  console.log(`Processing ${files.length} images...`);
  for (const f of files) {
    const p = join(IMG_DIR, f);
    await processImage(p);
    console.log(`  ${f}  done`);
  }
  console.log('All done.');
}

main().catch(e => { console.error(e); process.exit(1); });
